### PR TITLE
Makefile proposal

### PR DESCRIPTION
**Pull request by scruffystuffs**
# Overview

This is a 15-min attempt at producing a convenience makefile.  This is just to see how people feel about introducing something like this into the codebase.  If we all like it, we can merge it, but if not, we can throw it away.

This can greatly help us when introducing linting and formatting into CI, although it's not necessary.

## Acceptance criteria

* There's a makefile
* It works for macOS and linux

## Checklist

- [x] I added tests for this PR's change.
- [x] I linted and formatted (via `haskell-language-server`) any files I touched in this PR.
- [x] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [x] I updated `Changelog.md`. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [x] I linked this PR to any referenced GitHub issues.

### CODE REVIEW

File: .gitignore
Comment (liftM) [inline .gitignore:15]: [nit] Consistent whitespace between sections.

```suggestion

# Executables
```

File: .gitignore
Comment (liftM) [inline .gitignore:13]: Why `sandbox`? Are multiple people using this convention? I'm personally already using `.scratch`, with my own folder structure for test projects.

File: Makefile
Comment (liftM) [inline Makefile:22]: Like above, I'd remove this unless other folks are already using `./sandbox` as a convention.

File: Makefile
Comment (liftM) [inline Makefile:17]: I'd just call this `make install`, not `make fossa` or `make install-local`.

File: Makefile
Comment (liftM) [inline Makefile:26]: This is interesting.

I'm not sure this is _useful_ in general. You can't replay a customer debug log unless you use the same folder name, and there's no point in replaying your own `./sandbox` run because you could just re-run `make build && make sandbox` since you'd have to have the project files locally to generate the record in the first place.

Unless I'm missing something, this doesn't seem useful. I'd remove it.

File: Makefile
Comment (liftM) [inline Makefile:36]: I'd combine this into `make lint`, so there's a single command that you run to make sure everything is okay before a merge.

File: Makefile
Comment (liftM) [inline Makefile:38]: > This parses wrong sometimes

That's alarming. When does this happen?

**Review by liftM — CHANGES_REQUESTED** [review]
I like it, but I would remove a couple of the less general targets.

File: Makefile
Comment (scruffystuffs) [reply Makefile:38]: When extensions are not activated.  For now it seems to be OK, and it helps that we don't use default extensions in cabal unless they're no-brainers.  But it can happen, and HLS prevents it.

File: Makefile
Comment (scruffystuffs) [reply Makefile:36]: I purposely left these separate so that CI can run the two processes independently, that way if both fail, you get all the info you need to fix it.

File: Makefile
Comment (liftM) [reply Makefile:38]: We can configure hlint extensions via flags or `.hlint.yaml` ([doc](https://github.com/ndmitchell/hlint#language-extensions)). Let's do that instead of leaving this spooky comment in.

File: Makefile
Comment (scruffystuffs) [reply Makefile:36]: We can combine them into a `check` target, but keeping them separate is intentional.

Comment (liftM) [reply Makefile:36]: `make check` SGTM. In that case, I would still rename `fmt-ck` to `check-fmt` (or something less terse).

File: Makefile
Comment (scruffystuffs) [reply Makefile:17]: It's intentionally given the same name as the file it produces.  `make` will not rebuild it if it already exists, which saves some time.

File: .gitignore
Comment (scruffystuffs) [reply .gitignore:13]: Just an idea.  I liked the idea of sandbox because of the pre-existing connotation of it being isolated.

File: .github/pull_request_template.md
Comment (liftM) [inline .github/pull_request_template.md:25]: Will this shortened form work for repos across the `fossas` org?

File: Makefile
Comment (scruffystuffs) [reply Makefile:26]: Yeah, I forgot about the working directory part when writing this.

File: .gitignore
Comment (liftM) [reply .gitignore:13]: Given that we already have `make fossa`, I don't encoding the `sandbox` convention adds much value. I'd personally remove references to `sandbox`, but if you want to start this as a convention and drive any tooling around it, I'm okay with that too.

File: .github/pull_request_template.md
Comment (scruffystuffs) [reply .github/pull_request_template.md:25]: Yep.  Edit this comment and see how fossas/team-analysis#1 is rendered.  I've been doing it on PR's for a few months now.  It actually works across GH, not just within an org.

Comment (scruffystuffs) [reply .github/pull_request_template.md:25]: I think it should be `org/repo#number` in the example text, though

File: .gitignore
Comment (scruffystuffs) [reply .gitignore:13]: We need somewhere that git will ignore, it's really annoying to have it always tell me there are untracked files.  I'll leave it, but I'll also make it more modular so we can decide on some other name if we want.  And this only affects us, so I'm not worried about committing to anything.

Comment (scruffystuffs) [reply .gitignore:13]: If I develop other tooling it will be modular with that name as well.

Comment (liftM) [reply .gitignore:13]: > We need somewhere that git will ignore

I don't think picking an arbitrary folder name is the right solution for this. Developers should configure their personal preferences. You can set up [personal repository exclusion settings](https://stackoverflow.com/questions/25912441/can-i-add-a-gitignore-file-just-for-me-that-overrides-the-project-gitignore), although I use a [personal global gitignore](https://stackoverflow.com/questions/5724455/can-i-make-a-user-specific-gitignore-file).

Comment (scruffystuffs) [reply .gitignore:13]: I also avoid doing either of those exclusion methods, since it can be difficult to track down why something is excluded if it's not right there in the project.  Many people don't know about either of those features well enough to use them without googling it (though they are pretty simple once you look) and providing an obvious place to put test projects seems a lot simpler to me.  

Is picking an arbitrary name really so bad?  I don't see how this would have such a negative effect.  We could just remove it if no one else uses it.

Comment (liftM) [reply .gitignore:13]: > Is picking an arbitrary name really so bad? I don't see how this would have such a negative effect. We could just remove it if no one else uses it.

I just think that this behavior ("hey there's a folder named `sandbox` that for some reason is ignored") is surprising, and that it provides little value given that developers can really easily set up personal gitignores. The articles I linked were the top 2 results for "personal gitignore".

I'm not going to fight hard to block this, but I don't think it's the right tool for this problem.

**Review by liftM — APPROVED** [review]

File: Makefile
Comment (liftM) [reply Makefile]: Before you merge - did you add these flags @scruffystuffs? I'm not seeing an `.hlint.yaml` either.

Comment (scruffystuffs) [reply Makefile]: That's already in the repo from an earlier PR.  See #258

**Comment by scruffystuffs** [comment]
@spatten @cnr I'm going to merge this, we can remove if it's not the right approach.

**Pull request merged**
